### PR TITLE
add use-primitive-type rule

### LINKED ISSUE #385: Rule: Wrapper objects should not be used for primitive types (RSPEC-1533)
[RSPEC-1533](https://jira.sonarsource.com/browse/RSPEC-1533)

```
function foo(name: String, secondName: string) { // "string" should be used instead of "String"
}
```

### PR DESCRIPTION
**Pull request by stas-vilchik**
Fixes #385

### CODE REVIEW

File: sonarts-core/src/rules/usePrimitiveTypeRule.ts [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+  applyWithProgram(sourceFile: ts.SourceFile, program: ts.Program): tslint.RuleFailure[] {
+    return this.applyWithWalker(new Walker(sourceFile, this.getOptions(), program));
+  }
+}
+
+class Walker extends tslint.ProgramAwareRuleWalker {
+  protected visitNode(node: ts.Node) {
+    if (ts.isNewExpression(node)) {
+      this.checkNewExpression(node);
+    } else if (isFunctionLikeDeclaration(node)) {
```
Comment (vilchik-elena) [inline sonarts-core/src/rules/usePrimitiveTypeRule.ts]: @stas-vilchik why are we limiting ourself to only parameters and var decl type? We are missing case like this for example `function foo(p: String | number){}`. Let's visit directly TypeNode

File: sonarts-core/src/rules/usePrimitiveTypeRule.ts [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+
+  private checkNewExpression(node: ts.NewExpression) {
+    if (ts.isIdentifier(node.expression) && this.isPrimitiveWrapper(node.expression.text)) {
+      this.addFailureAtNode(node, `Remove this use of '${node.expression.text}' constructor.`);
+    }
+  }
+
+  private checkTypeNode(typeNode?: ts.TypeNode) {
+    if (typeNode) {
+      const type = this.getTypeChecker().getTypeFromTypeNode(typeNode);
```
Comment (vilchik-elena) [inline sonarts-core/src/rules/usePrimitiveTypeRule.ts]: @stas-vilchik why type checker here? can't we just `isPrimitiveWrapper(typeNode.text)`?

File: sonarts-sq-plugin/sonar-typescript-plugin/src/main/resources/org/sonar/l10n/typescript/rules/typescript/S1533.json [hunk 3]
```diff
@@ -9,7 +9,5 @@
   "tags": [
     "pitfall"
   ],
-  "defaultSeverity": "Minor",
-  "ruleSpecification": "RSPEC-1533",
-  "sqKey": "S1533"
+  "defaultSeverity": "Minor"
```
Comment (vilchik-elena) [inline sonarts-sq-plugin/sonar-typescript-plugin/src/main/resources/org/sonar/l10n/typescript/rules/typescript/S1533.json]: @stas-vilchik are you using new rule-api?

File: sonarts-core/src/rules/usePrimitiveTypeRule.ts (hunk 2 above)
Comment (stas-vilchik) [reply sonarts-core/src/rules/usePrimitiveTypeRule.ts]: in this case we don't check the symbol and start generating issues on
```typescript
type String = number;
function stringTypeAlias(x: String) {}
```

Comment (vilchik-elena) [reply sonarts-core/src/rules/usePrimitiveTypeRule.ts]: still this logic doesn't save us from this FP https://github.com/SonarSource/SonarTS/pull/398/files#diff-3848ed87910db18c6b8f29a425190b41R80

IMO it's better to be consistent in FPs :) and have them in both places

File: sonarts-core/src/rules/usePrimitiveTypeRule.ts [hunk 4]
```diff
@@ -46,34 +45,25 @@ export class Rule extends tslint.Rules.TypedRule {
 
 class Walker extends tslint.ProgramAwareRuleWalker {
   protected visitNode(node: ts.Node) {
-    if (ts.isNewExpression(node)) {
-      this.checkNewExpression(node);
-    } else if (isFunctionLikeDeclaration(node)) {
-      node.parameters.forEach(parameter => this.checkTypeNode(parameter.type));
-    } else if (ts.isVariableDeclaration(node) || ts.isPropertyDeclaration(node)) {
-      this.checkTypeNode(node.type);
+    // TODO create and use `visitTypeNode` in visitor
```
Comment (vilchik-elena) [inline sonarts-core/src/rules/usePrimitiveTypeRule.ts]: @stas-vilchik rebase on master and make this change

**Pull request merged**